File: Wikipedia/Code/AccountCreationViewController.m
#import "AccountCreationViewController.h"
#import "SessionSingleton.h"
#import "UIScrollView+ScrollSubviewToLocation.h"
#import "LoginViewController.h"
#import "PreviewAndSaveViewController.h"
#import "UIBarButtonItem+WMFButtonConvenience.h"
#import "UIViewController+WMFChildViewController.h"
#import "UIViewController+WMFStoryboardUtilities.h"
#import "SectionEditorViewController.h"
#import "Wikipedia-Swift.h"
#import "PaddedLabel.h"
#import "MWKLanguageLinkController.h"
#import "WMFAuthenticationManager.h"
#import <SDWebImage/SDWebImage.h>

@import BlocksKitUIKitExtensions;

@interface AccountCreationViewController ()

@property (strong, nonatomic) CaptchaViewController *captchaViewController;
@property (weak, nonatomic) IBOutlet UIView *captchaContainer;
@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;
@property (nonatomic) BOOL showCaptchaContainer;
@property (strong, nonatomic) NSString *captchaId;
@property (strong, nonatomic) NSURL *captchaUrl;
@property (weak, nonatomic) IBOutlet PaddedLabel *loginButton;
@property (weak, nonatomic) IBOutlet PaddedLabel *titleLabel;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *usernameUnderlineHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *passwordUnderlineHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *passwordConfirmUnderlineHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *emailUnderlineHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *spaceBeneathCaptchaContainer;

@property (weak, nonatomic) IBOutlet UIView *createAccountContainerView;

@property (strong, nonatomic) UIBarButtonItem *rightButton;

@end

@implementation AccountCreationViewController

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    // Disable horizontal scrolling.
    scrollView.contentOffset = CGPointMake(0.0, scrollView.contentOffset.y);
}

- (void)updateViewConstraints {
    [self adjustScrollLimitForCaptchaVisiblity];

    [super updateViewConstraints];
}

- (void)adjustScrollLimitForCaptchaVisiblity {
    // Reminder: spaceBeneathCaptchaContainer constraint is space *below* captcha container - that's why below
    // for the show case we don't have to "convertPoint".
    self.spaceBeneathCaptchaContainer.constant =
        (self.showCaptchaContainer)
            ? (self.view.frame.size.height - (self.captchaContainer.frame.size.height / 2))
            : (self.view.frame.size.height - [self.loginButton convertPoint:CGPointZero toView:self.scrollView].y);
    ;
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    // Ensure adjustScrollLimitForCaptchaVisiblity gets called again after rotating.
    [self.view setNeedsUpdateConstraints];
}

- (void)viewDidLoad {
    [super viewDidLoad];

    @weakify(self)
        UIBarButtonItem *xButton = [UIBarButtonItem wmf_buttonType:WMFButtonTypeX
                                                           handler:^(id sender) {
                                                               @strongify(self) if (self.showCaptchaContainer) {
                                                                   self.showCaptchaContainer = NO;
                                                               }
                                                               else {
                                                                   [self dismissViewControllerAnimated:YES completion:nil];
                                                               }
                                                           }];
    self.navigationItem.leftBarButtonItems = @[xButton];

    self.rightButton = [[UIBarButtonItem alloc] bk_initWithTitle:MWLocalizedString(@"button-next", nil)
                                                           style:UIBarButtonItemStylePlain
                                                         handler:^(id sender) {
                                                             @strongify(self)

                                                                 [[WMFAlertManager sharedInstance] showAlert:MWLocalizedString(@"account-creation-saving", nil)
                                                                                                      sticky:YES
                                                                                       dismissPreviousAlerts:YES
                                                                                                 tapCallBack:NULL];

                                                             [self save];
                                                         }];
    self.navigationItem.rightBarButtonItem = self.rightButton;

    self.titleLabel.font = [UIFont boldSystemFontOfSize:23.0f];
    self.usernameField.font = [UIFont boldSystemFontOfSize:18.0f];
    self.passwordField.font = [UIFont boldSystemFontOfSize:18.0f];
    self.passwordRepeatField.font = [UIFont boldSystemFontOfSize:18.0f];
    self.emailField.font = [UIFont boldSystemFontOfSize:18.0f];
    self.loginButton.font = [UIFont boldSystemFontOfSize:14.0f];

    self.captchaId = @"";
    self.captchaUrl = nil;
    self.scrollView.delegate = self;

    [self.usernameField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];

    [self.passwordField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];

    [self.passwordRepeatField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];

    [self.emailField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];

    self.captchaContainer.alpha = 0;

    self.usernameField.attributedPlaceholder =
        [self getAttributedPlaceholderForString:MWLocalizedString(@"account-creation-username-placeholder-text", nil)];

    self.passwordField.attributedPlaceholder =
        [self getAttributedPlaceholderForString:MWLocalizedString(@"account-creation-password-placeholder-text", nil)];

    self.passwordRepeatField.attributedPlaceholder =
        [self getAttributedPlaceholderForString:MWLocalizedString(@"account-creation-password-confirm-placeholder-text", nil)];

    self.emailField.attributedPlaceholder =
        [self getAttributedPlaceholderForString:MWLocalizedString(@"account-creation-email-placeholder-text", nil)];

    if ([self.scrollView respondsToSelector:@selector(keyboardDismissMode)]) {
        self.scrollView.keyboardDismissMode = UIScrollViewKeyboardDismissModeInteractive;
    }

    self.usernameUnderlineHeight.constant = 1.0f / [UIScreen mainScreen].scale;
    self.passwordUnderlineHeight.constant = self.usernameUnderlineHeight.constant;
    self.passwordConfirmUnderlineHeight.constant = self.usernameUnderlineHeight.constant;
    self.emailUnderlineHeight.constant = self.usernameUnderlineHeight.constant;

    self.loginButton.textColor = [UIColor wmf_blueTintColor];
    self.loginButton.padding = UIEdgeInsetsMake(10, 10, 10, 10);
    self.loginButton.text = MWLocalizedString(@"account-creation-login", nil);
    self.loginButton.userInteractionEnabled = YES;
    [self.loginButton addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(loginButtonPushed:)]];

    self.titleLabel.text = MWLocalizedString(@"navbar-title-mode-create-account", nil);

    self.usernameField.textAlignment = NSTextAlignmentNatural;
    self.passwordField.textAlignment = NSTextAlignmentNatural;
    self.passwordRepeatField.textAlignment = NSTextAlignmentNatural;
    self.emailField.textAlignment = NSTextAlignmentNatural;
}

- (void)loginButtonPushed:(UITapGestureRecognizer *)recognizer {
    if (recognizer.state == UIGestureRecognizerStateEnded) {
        UIViewController *presenter = self.presentingViewController;

        [self dismissViewControllerAnimated:YES
                                 completion:^{
                                     UINavigationController *nc = [[UINavigationController alloc] initWithRootViewController:[LoginViewController wmf_initialViewControllerFromClassStoryboard]];

                                     [presenter presentViewController:nc animated:YES completion:nil];
                                 }];
    }
}

- (NSAttributedString *)getAttributedPlaceholderForString:(NSString *)string {
    return [[NSMutableAttributedString alloc] initWithString:string
                                                  attributes:@{
                                                      NSForegroundColorAttributeName: [UIColor lightGrayColor]
                                                  }];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    self.captchaViewController = [CaptchaViewController wmf_initialViewControllerFromClassStoryboard];
    [self wmf_addChildController:self.captchaViewController andConstrainToEdgesOfContainerView:self.captchaContainer];

    self.showCaptchaContainer = NO;

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(textFieldDidChange:)
                                                 name:@"UITextFieldTextDidChangeNotification"
                                               object:self.captchaViewController.captchaTextBox];

    [self enableProgressiveButton:NO];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self.usernameField becomeFirstResponder];
}

- (void)textFieldDidChange:(id)sender {
    BOOL shouldHighlight = ((self.usernameField.text.length > 0) &&
                            (self.passwordField.text.length > 0) &&
                            (self.passwordRepeatField.text.length > 0) &&
                            //(self.emailField.text.length > 0) &&
                            [self.passwordField.text isEqualToString:self.passwordRepeatField.text])
                               ? YES
                               : NO;

    // Override shouldHighlight if the text changed was the captcha field.
    if ([sender isKindOfClass:[NSNotification class]]) {
        NSNotification *notification = (NSNotification *)sender;
        if (notification.object == self.captchaViewController.captchaTextBox) {
            NSString *trimmedCaptchaText =
                [self.captchaViewController.captchaTextBox.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
            shouldHighlight = (trimmedCaptchaText.length > 0) ? YES : NO;
        }
    }

    [self enableProgressiveButton:shouldHighlight];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == self.usernameField) {
        [self.passwordField becomeFirstResponder];
    } else if (textField == self.passwordField) {
        [self.passwordRepeatField becomeFirstResponder];
    } else if (textField == self.passwordRepeatField) {
        [self.emailField becomeFirstResponder];
    } else {
        NSAssert(((textField == self.emailField) || (textField == self.captchaViewController.captchaTextBox)),
                 @"Received -textFieldShouldReturn for unexpected text field: %@", textField);
        [self save];
    }
    return YES;
}

- (void)enableProgressiveButton:(BOOL)enabled {
    self.navigationItem.rightBarButtonItem.enabled = enabled;
}

- (void)prepopulateTextFieldsForDebugging {
#if DEBUG
    self.usernameField.text = @"acct_creation_test_010";
    self.passwordField.text = @"";
    self.passwordRepeatField.text = @"";
    self.emailField.text = @"mhurd@wikimedia.org";
#endif
}

- (void)viewWillDisappear:(BOOL)animated {
    [self enableProgressiveButton:NO];

    [[WMFAlertManager sharedInstance] dismissAlert];

    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:@"UITextFieldTextDidChangeNotification"
                                                  object:self.captchaViewController.captchaTextBox];

    [super viewWillDisappear:animated];
}

- (void)setShowCaptchaContainer:(BOOL)showCaptchaContainer {
    _showCaptchaContainer = showCaptchaContainer;

    self.rightButton.title = showCaptchaContainer ? MWLocalizedString(@"button-done", nil) : MWLocalizedString(@"button-next", nil);

    CGFloat duration = 0.5;

    [self.view setNeedsUpdateConstraints];

    if (showCaptchaContainer) {
        [self.captchaViewController.captchaTextBox performSelector:@selector(becomeFirstResponder)
                                                        withObject:nil
                                                        afterDelay:0.4f];
        [self.funnel logCaptchaShown];
        dispatch_async(dispatch_get_main_queue(), ^(void) {
            [UIView animateWithDuration:duration
                animations:^{
                    self.captchaContainer.alpha = 1;
                    [self.scrollView scrollSubViewToTop:self.captchaContainer animated:NO];
                }
                completion:^(BOOL done) {
                    [self enableProgressiveButton:NO];
                }];
        });
    } else {
        dispatch_async(dispatch_get_main_queue(), ^(void) {
            [[WMFAlertManager sharedInstance] dismissAlert];
            [UIView animateWithDuration:duration
                animations:^{
                    self.captchaContainer.alpha = 0;
                    [self.scrollView setContentOffset:CGPointZero animated:NO];
                }
                completion:^(BOOL done) {
                    self.captchaViewController.captchaTextBox.text = @"";
                    self.captchaViewController.captchaImageView.image = nil;
                    // Pretent a text field changed so the progressive button state gets updated.
                    [self textFieldDidChange:nil];
                }];
        });
    }
}

- (void)setCaptchaUrl:(NSURL *)captchaUrl {
    if (![_captchaUrl isEqual:captchaUrl]) {
        _captchaUrl = captchaUrl;
        if (captchaUrl) {
            [self refreshCaptchaImage];
        }
    }
}

- (void)refreshCaptchaImage {
    self.captchaViewController.captchaTextBox.text = @"";
    [self.captchaViewController.captchaImageView sd_setImageWithURL:self.captchaUrl];
    self.showCaptchaContainer = YES;
}

- (void)reloadCaptchaPushed:(id)sender {
    self.captchaViewController.captchaTextBox.text = @"";

    [[WMFAlertManager sharedInstance] showAlert:MWLocalizedString(@"account-creation-captcha-obtaining", nil) sticky:NO dismissPreviousAlerts:YES tapCallBack:NULL];

    [self save]; //this resarts the token process which gets us another captcha
}

- (void)login {
    [[WMFAlertManager sharedInstance] showAlert:MWLocalizedString(@"account-creation-logging-in", nil) sticky:YES dismissPreviousAlerts:YES tapCallBack:NULL];
    [[WMFAuthenticationManager sharedInstance] loginWithUsername:self.usernameField.text
        password:self.passwordField.text
        success:^{
            NSString *loggedInMessage = MWLocalizedString(@"main-menu-account-title-logged-in", nil);
            loggedInMessage = [loggedInMessage stringByReplacingOccurrencesOfString:@"$1"
                                                                         withString:self.usernameField.text];
            [[WMFAlertManager sharedInstance] showSuccessAlert:loggedInMessage sticky:NO dismissPreviousAlerts:YES tapCallBack:NULL];
            [self dismissViewControllerAnimated:YES completion:nil];
        }
        failure:^(NSError *error) {
            [self enableProgressiveButton:YES];
        }];
}

- (NSArray *)requiredInputFields {
    NSAssert([self isViewLoaded],
             @"This method is only intended to be called when view is loaded, since they'll all be nil otherwise");
    return @[self.usernameField, self.passwordField, self.passwordRepeatField];
}

- (BOOL)isPasswordConfirmationCorrect {
    return [self.passwordField.text isEqualToString:self.passwordRepeatField.text];
}

- (BOOL)areRequiredFieldsPopulated {
    return NSNotFound ==
           [[self requiredInputFields] indexOfObjectPassingTest:^BOOL(UITextField *field, NSUInteger idx, BOOL *stop) {
               if (field.text.length == 0) {
                   *stop = YES;
                   return YES;
               } else {
                   return NO;
               }
           }];
}

- (void)save {
    if (![self areRequiredFieldsPopulated]) {
        [[WMFAlertManager sharedInstance] showErrorAlertWithMessage:MWLocalizedString(@"account-creation-missing-fields", nil) sticky:YES dismissPreviousAlerts:YES tapCallBack:NULL];
        return;
    }

    // Verify passwords fields match.
    if (![self isPasswordConfirmationCorrect]) {
        [[WMFAlertManager sharedInstance] showErrorAlertWithMessage:MWLocalizedString(@"account-creation-passwords-mismatched", nil) sticky:YES dismissPreviousAlerts:YES tapCallBack:NULL];
        return;
    }

    if ([self.captchaViewController.captchaTextBox.text length] == 0) {
        [[WMFAuthenticationManager sharedInstance] getAccountCreationCaptchaWithUsername:self.usernameField.text
            password:self.passwordField.text
            email:self.emailField.text
            captcha:^(NSURL *_Nonnull captchaURL) {
                self.captchaUrl = captchaURL;

                [[WMFAlertManager sharedInstance] showWarningAlert:MWLocalizedString(@"account-creation-captcha-required", nil) sticky:NO dismissPreviousAlerts:YES tapCallBack:NULL];
            }
            failure:^(NSError *error) {
                [[WMFAlertManager sharedInstance] showErrorAlert:error sticky:YES dismissPreviousAlerts:YES tapCallBack:NULL];
                [self.funnel logError:error.localizedDescription];
            }];
    } else {
        [[WMFAuthenticationManager sharedInstance] createAccountWithCaptchaText:self.captchaViewController.captchaTextBox.text
            success:^{
                NSString *loggedInMessage = MWLocalizedString(@"main-menu-account-title-logged-in", nil);
                loggedInMessage = [loggedInMessage stringByReplacingOccurrencesOfString:@"$1"
                                                                             withString:self.usernameField.text];
                [[WMFAlertManager sharedInstance] showSuccessAlert:loggedInMessage sticky:NO dismissPreviousAlerts:YES tapCallBack:NULL];
                [self dismissViewControllerAnimated:YES completion:nil];
            }
            captcha:^(NSURL *_Nonnull captchaURL) {
                self.captchaUrl = captchaURL;

                [[WMFAlertManager sharedInstance] showWarningAlert:MWLocalizedString(@"account-creation-captcha-required", nil) sticky:NO dismissPreviousAlerts:YES tapCallBack:NULL];
            }
            failure:^(NSError *error) {
                [[WMFAlertManager sharedInstance] showErrorAlert:error sticky:YES dismissPreviousAlerts:YES tapCallBack:NULL];
                [self.funnel logError:error.localizedDescription];
                [self enableProgressiveButton:YES];
            }];
    }
}

@end

File: Wikipedia/Code/AccountCreationViewController.swift

import UIKit

class AccountCreationViewController: UIViewController, CaptchaViewControllerRefresh, UITextFieldDelegate, UIScrollViewDelegate {
    @IBOutlet var usernameField: UITextField!
    @IBOutlet var passwordField: UITextField!
    @IBOutlet var passwordRepeatField: UITextField!
    @IBOutlet var emailField: UITextField!
    @IBOutlet var captchaContainer: UIView!
    @IBOutlet var scrollView: UIScrollView!
    @IBOutlet var loginButton: UILabel!
    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var usernameUnderlineHeight: NSLayoutConstraint!
    @IBOutlet var passwordUnderlineHeight: NSLayoutConstraint!
    @IBOutlet var passwordConfirmUnderlineHeight: NSLayoutConstraint!
    @IBOutlet var emailUnderlineHeight: NSLayoutConstraint!
    @IBOutlet var spaceBeneathCaptchaContainer: NSLayoutConstraint!
    @IBOutlet var createAccountContainerView: UIView!

    fileprivate var captchaId: NSString? = ""
    fileprivate var rightButton: UIBarButtonItem?
    public var funnel: CreateAccountFunnel?
    fileprivate var captchaViewController: CaptchaViewController?
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        // Disable horizontal scrolling.
        scrollView.contentOffset = CGPoint(x: 0.0, y: scrollView.contentOffset.y)
    }
    
    fileprivate func adjustScrollLimitForCaptchaVisiblity() {
        // Reminder: spaceBeneathCaptchaContainer constraint is space *below* captcha container -
        // that's why below for the show case we don't have to "convertPoint".
        spaceBeneathCaptchaContainer.constant = (showCaptchaContainer)
            ? (view.frame.size.height - (captchaContainer.frame.size.height / 2))
            : (view.frame.size.height - loginButton.convert(CGPoint.zero, to:scrollView).y)
    }

    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        // Ensure adjustScrollLimitForCaptchaVisiblity gets called again after rotating.
        view.setNeedsUpdateConstraints()
    }
    
    func didTapClose(_ tap: UITapGestureRecognizer) {
        if (showCaptchaContainer) {
            showCaptchaContainer = false
        } else {
            self.dismiss(animated: true, completion: nil)
        }
    }

    func loginButtonPushed(_ recognizer: UITapGestureRecognizer) {
        guard
            recognizer.state == .ended,
            let presenter = self.presentingViewController else {
                return
        }
        dismiss(animated: true, completion: {
            let navigationController = UINavigationController.init(rootViewController: LoginViewController.wmf_initialViewControllerFromClassStoryboard()!)
            presenter.present(navigationController, animated: true, completion: nil)
        })
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named:"close"), style: .plain, target:self, action:#selector(self.didTapClose(_:)))

        rightButton = UIBarButtonItem(title: localizedStringForKeyFallingBackOnEnglish("button-next"), style: .plain, target: self, action: #selector(self.doneButtonPushed(_:)))
        
        navigationItem.rightBarButtonItem = rightButton
        
        scrollView.delegate = self
        
        usernameField.addTarget(self, action: #selector(self.textFieldDidChange(_:)), for: .editingChanged)
        passwordField.addTarget(self, action: #selector(self.textFieldDidChange(_:)), for: .editingChanged)
        passwordRepeatField.addTarget(self, action: #selector(self.textFieldDidChange(_:)), for: .editingChanged)
        emailField.addTarget(self, action: #selector(self.textFieldDidChange(_:)), for: .editingChanged)

        captchaContainer.alpha = 0
        
        usernameField.placeholder = localizedStringForKeyFallingBackOnEnglish("account-creation-username-placeholder-text")
        passwordField.placeholder = localizedStringForKeyFallingBackOnEnglish("account-creation-password-placeholder-text")
        passwordRepeatField.placeholder = localizedStringForKeyFallingBackOnEnglish("account-creation-password-confirm-placeholder-text")
        emailField.placeholder = localizedStringForKeyFallingBackOnEnglish("account-creation-email-placeholder-text")

        scrollView.keyboardDismissMode = .interactive
        
        usernameUnderlineHeight.constant = 1.0 / UIScreen.main.scale
        passwordUnderlineHeight.constant = usernameUnderlineHeight.constant;
        passwordConfirmUnderlineHeight.constant = usernameUnderlineHeight.constant;
        emailUnderlineHeight.constant = usernameUnderlineHeight.constant;

        loginButton.textColor = UIColor.wmf_blueTint()
        loginButton.text = localizedStringForKeyFallingBackOnEnglish("account-creation-login")
        loginButton.isUserInteractionEnabled = true
        
        loginButton.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(self.loginButtonPushed(_:))))
        titleLabel.text = localizedStringForKeyFallingBackOnEnglish("navbar-title-mode-create-account")
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        captchaViewController = CaptchaViewController.wmf_initialViewControllerFromClassStoryboard()
        wmf_addChildController(captchaViewController, andConstrainToEdgesOfContainerView: captchaContainer)
        showCaptchaContainer = false
        NotificationCenter.default.addObserver(self, selector: #selector(self.textFieldDidChange(_:)), name: NSNotification.Name.UITextFieldTextDidChange, object: captchaViewController?.captchaTextBox)
        enableProgressiveButton(false)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        usernameField.becomeFirstResponder()
    }

    func textFieldDidChange(_ sender: Any?) {
        var shouldHighlight = ((usernameField.text!.characters.count > 0 && passwordField.text!.characters.count > 0 && passwordRepeatField.text!.characters.count > 0 && passwordField.text == passwordRepeatField.text))
        // Override shouldHighlight if the text changed was the captcha field.
        if let notification = sender as? Notification {
            if notification.object as AnyObject? === captchaViewController?.captchaTextBox {
                let trimmedCaptchaText = captchaViewController?.captchaTextBox.text?.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
                shouldHighlight = ((trimmedCaptchaText?.characters.count)! > 0)
            }
        }
        enableProgressiveButton(shouldHighlight)
    }

    public func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if (textField == usernameField) {
            passwordField.becomeFirstResponder()
        } else if (textField == passwordField) {
            passwordRepeatField.becomeFirstResponder()
        } else if (textField == passwordRepeatField) {
            emailField.becomeFirstResponder()
        } else {
            assert(((textField == emailField) || (textField == captchaViewController?.captchaTextBox)), "Received -textFieldShouldReturn for unexpected text field: \(textField)")
            save()
        }
        return true
    }

    fileprivate func enableProgressiveButton(_ enabled: Bool) {
        navigationItem.rightBarButtonItem?.isEnabled = enabled
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        enableProgressiveButton(false)
        WMFAlertManager.sharedInstance.dismissAlert()
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UITextFieldTextDidChange, object: captchaViewController?.captchaTextBox)
        super.viewWillDisappear(animated)
    }
    
    fileprivate var showCaptchaContainer: Bool = false {
        didSet {
            rightButton?.title = showCaptchaContainer ? localizedStringForKeyFallingBackOnEnglish("button-done") : localizedStringForKeyFallingBackOnEnglish("button-next")
            let duration: TimeInterval = 0.5
            view.setNeedsUpdateConstraints()
            
            if showCaptchaContainer {
                captchaViewController?.captchaTextBox.perform(#selector(becomeFirstResponder), with: nil, afterDelay: 0.4)
                funnel?.logCaptchaShown()
                DispatchQueue.main.async(execute: {
                    UIView.animate(withDuration: duration, animations: {
                        self.captchaContainer.alpha = 1
                        self.scrollView.scrollSubView(toTop: self.captchaContainer, animated:false)
                    }, completion: {(completed: Bool) -> Void in
                        self.enableProgressiveButton(false)
                    })
                })
            }else{
                DispatchQueue.main.async(execute: {
                    WMFAlertManager.sharedInstance.dismissAlert()
                    UIView.animate(withDuration: duration, animations: {
                        self.captchaContainer.alpha = 0
                        self.scrollView.setContentOffset(CGPoint.zero, animated: false)
                    }, completion: {(completed: Bool) -> Void in
                        self.captchaViewController?.captchaTextBox.text = ""
                        self.captchaViewController?.captchaImageView.image = nil
                        // Pretent a text field changed so the progressive button state gets updated.
                        self.textFieldDidChange(nil)
                    })
                })
            }
        }
    }
    
    fileprivate var captchaURL: URL? {
        didSet {
            guard captchaURL != nil else {
                return;
            }
            refreshCaptchaImage()
        }
    }
    
    fileprivate func refreshCaptchaImage() {
        captchaViewController?.captchaTextBox.text = ""
        captchaViewController?.captchaImageView.sd_setImage(with: captchaURL)
        showCaptchaContainer = true
    }

    func reloadCaptchaPushed(_ sender: AnyObject) {
        captchaViewController?.captchaTextBox.text = ""
        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-captcha-obtaining"), sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
        save() //this restarts the token process which gets us another captcha
    }
    
    fileprivate func login() {
        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-logging-in"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
        WMFAuthenticationManager.sharedInstance().login(
            withUsername: usernameField.text!,
            password: passwordField.text!,
            retypePassword: nil, 
            success: {
                let loggedInMessage = localizedStringForKeyFallingBackOnEnglish("main-menu-account-title-logged-in").replacingOccurrences(of: "$1", with: self.usernameField.text!)
                WMFAlertManager.sharedInstance.showSuccessAlert(loggedInMessage, sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
                self.dismiss(animated: true, completion: nil)
        }, failure: { error in
            self.enableProgressiveButton(true)
            // WMFAlertManager.sharedInstance.showErrorAlert(error as NSError, sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
        })
    }
    
    fileprivate func requiredInputFields() -> [UITextField] {
        assert(isViewLoaded, "This method is only intended to be called when view is loaded, since they'll all be nil otherwise")
        return [usernameField, passwordField, passwordRepeatField]
    }

    fileprivate func isPasswordConfirmationCorrect() -> Bool {
        return passwordField.text == passwordRepeatField.text
    }
    
    fileprivate func areRequiredFieldsPopulated() -> Bool {
        let firstRequiredFieldWithNoText = requiredInputFields().first(where:{ $0.text?.characters.count == 0 })
        return firstRequiredFieldWithNoText == nil
    }
    
    func doneButtonPushed(_ tap: UITapGestureRecognizer) {
        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-saving"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
        save()
    }
    
    override func updateViewConstraints() {
        adjustScrollLimitForCaptchaVisiblity()
        super.updateViewConstraints()
    }

    fileprivate func save() {
        guard areRequiredFieldsPopulated() == true else {
            WMFAlertManager.sharedInstance.showErrorAlertWithMessage(localizedStringForKeyFallingBackOnEnglish("account-creation-missing-fields"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
            return
        }

        guard isPasswordConfirmationCorrect() == true else {
            WMFAlertManager.sharedInstance.showErrorAlertWithMessage(localizedStringForKeyFallingBackOnEnglish("account-creation-passwords-mismatched"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
            return
        }

        guard let captcha = captchaViewController?.captchaTextBox.text, captcha.characters.count > 0 else {
            getCaptcha()
            return
        }
        createAccount(withCaptcha: captcha)
    }
    
    fileprivate func getCaptcha() {
        WMFAuthenticationManager.sharedInstance().getAccountCreationCaptcha(withUsername: usernameField.text!, password: passwordField.text!, email: emailField.text!, captcha: {captchaURL in
            self.captchaURL = captchaURL
            WMFAlertManager.sharedInstance.showWarningAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-captcha-required"), sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
        }, failure: {error in
            WMFAlertManager.sharedInstance.showErrorAlert(error as NSError, sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
            self.funnel?.logError(error.localizedDescription)
        })
    }
    
    fileprivate func createAccount(withCaptcha captcha:String) {
        WMFAuthenticationManager.sharedInstance().createAccount(withCaptchaText: captcha, success: {
            let loggedInMessage = localizedStringForKeyFallingBackOnEnglish("main-menu-account-title-logged-in").replacingOccurrences(of: "$1", with: self.usernameField.text!)
            WMFAlertManager.sharedInstance.showSuccessAlert(loggedInMessage, sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
            self.dismiss(animated: true, completion: nil)
        }, captcha: {captchaURL in
            self.captchaURL = captchaURL
            WMFAlertManager.sharedInstance.showWarningAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-captcha-retry"), sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
        }, failure: {error in
            WMFAlertManager.sharedInstance.showErrorAlert(error as NSError, sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
            self.funnel?.logError(error.localizedDescription)
            self.enableProgressiveButton(true)
        })
    }
}
